src/hardware/voodoo.cpp

```diff
@@ -35,154 +38,159 @@
 
 
 class VOODOO;
-static VOODOO* voodoo_dev;
+static std::unique_ptr<VOODOO> voodoo_dev;
+
+constexpr uint32_t voodoo_lfb_mask = 0xffff0000;
+static uint32_t voodoo_current_lfb = (VOODOO_INITIAL_LFB & voodoo_lfb_mask);
 
-static uint32_t voodoo_current_lfb=(VOODOO_INITIAL_LFB&0xffff0000);
+enum class EmulationType {
+    None,
+    Software,
+    OpenGL
+};
+
+enum class CardType : int {
+    Type1 = 1
+};
 
+constexpr const char* NONE_STR = "false";
+constexpr const char* SOFTWARE_STR = "software";
+constexpr const char* OPENGL_STR = "opengl";
+constexpr const char* AUTO_STR = "auto";
 
-class VOODOO:public Module_base{
+class VOODOO : public Module_base {
 private:
-	Bits emulation_type;
-public:
-	VOODOO(Section* configuration):Module_base(configuration){
-		emulation_type=-1;
+    EmulationType emulation_type;
 
-		Section_prop * section=static_cast<Section_prop *>(configuration);
-		std::string voodoo_type_str(section->Get_string("voodoo_card"));
-		if (voodoo_type_str=="false") {
-			emulation_type=0;
-		} else if (voodoo_type_str=="software" || (voodoo_type_str=="auto" && !OpenGL_using())) {
-			emulation_type=1;
-#if C_OPENGL
-		} else if (voodoo_type_str=="opengl" || voodoo_type_str=="auto") {
-			emulation_type=2;
-#else
-		} else if (voodoo_type_str=="auto") {
-			emulation_type=1;
-#endif
-		} else {
-			emulation_type=0;
+public:
+    VOODOO(Section* configuration) : Module_base(configuration), emulation_type(EmulationType::None) {
+        UpdateConfiguration(configuration);
+    }
+	
+	EmulationType GetEmulationTypeFromString(const std::string& voodoo_type_str) {
+		static const std::unordered_map<std::string, EmulationType> str_to_emulation_type {
+			{NONE_STR, EmulationType::None},
+			{SOFTWARE_STR, EmulationType::Software},
+			{OPENGL_STR, C_OPENGL ? EmulationType::OpenGL : EmulationType::None},
+			{AUTO_STR, C_OPENGL ? EmulationType::OpenGL : EmulationType::Software}
+		};
+
+		auto it = str_to_emulation_type.find(voodoo_type_str);
+		if (it != str_to_emulation_type.end()) {
+			return it->second;
 		}
 
-		if (emulation_type<=0) return;
-
+		return !OpenGL_using() ? EmulationType::Software : EmulationType::None;
+	}
+	
+	void UpdateConfiguration(Section* configuration) {
+		Section_prop* section = static_cast<Section_prop*>(configuration);
+		std::string voodoo_type_str(section->Get_string("voodoo_card"));
```

Comment:
Can we figure out a better name than `voodoo_card`? We could append `card` to half many things in the config (e.g. almost all sound devices), but we don't 😄

Reply:
I am not opposed on changing `voodoo_card`, however, any suggestions besides just `card`? @kcgen @johnnovak

Reply:
> I'd say a `"voodoo"` section name should be fine.

I like that

Reply:
Yeah, so voodoo section, then rename voodoo_card to voodoo_model.

Reply:
@kcgen @johnnovak Changing it causes a bunch of issues on other files...

Reply:
Renaming a config param and introducing a new config section should not be too hard as there are plenty examples on how to do it in the codebase.

You will figure it out if you persevere; that's an important skill to have as a coder 😄

Reply:
I'm going to, like I said, I was noob and swapped everything to UpperCamelCase! I'm going to do it tonight!

Reply:
Ok, so thinking more about this, I think it would be the best to rename the current `voodoo_card` property to `voodoo_renderer` which can be `software` or `opengl`, and I'm inclined to drop `auto` as it doesn't seem to be too useful.

Then we need a `voodoo_model` property which now would only control whether Voodoo emulation is enabled or not. But maybe in the future we could add support for later cards too. So, for now possible values would be: `off`, and `v1`.

Then as I mentioned, the `3dfx` section should be renamed to `voodoo`.

`voodoo_memory` can stay, but with this name.

Reply:
@johnnovak @kcgen I've chatted with @joncampbell123 about this and he clarified that `voodoo_card` actually makes sense because the different renderers use different Voodoo1 versions